routes/verify.go

```diff
@@ -452,3 +484,380 @@ func (fes *APIServer) sendEmail(email *mail.SGMailV3) {
 func (fes *APIServer) IsConfiguredForSendgrid() bool {
 	return fes.Config.SendgridApiKey != ""
 }
+
+//
+// JUMIO
+//
+
+func (fes *APIServer) IsConfiguredForJumio() bool {
+	return fes.Config.JumioToken != ""  && fes.Config.JumioSecret != ""
+}
+
+type JumioInitRequest struct {
+	CustomerInternalReference string `json:"customerInternalReference"`
+	UserReference             string `json:"userReference"`
+	SuccessURL                string `json:"successUrl"`
+	ErrorURL                  string `json:"errorUrl"`
+}
+
+type JumioInitResponse struct {
+	RedirectURL          string `json:"redirectUrl"`
+	TransactionReference string `json:"transactionReference"`
+}
+
+type JumioBeginRequest struct {
+	PublicKey string
+	SuccessURL string
+	ErrorURL string
+	JWT       string
+}
+
+type JumioBeginResponse struct {
+	URL string
+}
+
+func (fes *APIServer) JumioBegin(ww http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := JumioBeginRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: Problem parsing request body: %v", err))
+		return
+	}
+
+	isValid, err := fes.ValidateJWT(requestData.PublicKey, requestData.JWT)
+	if !isValid {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: Invalid token: %v", err))
+		return
+	}
+
+	// Get UserMetadata from global state and check that user has not already been through Jumio Verification flow
+	userMetadata, err := fes.getUserMetadataFromGlobalState(requestData.PublicKey)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: Problem getting user metadata from global state: %v", err))
+		return
+	}
+
+	if userMetadata.JumioVerified {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: public key already went through jumio verification flow: %v", requestData.PublicKey))
+		return
+	}
+
+	if userMetadata.JumioFinishedTime > 0 && !userMetadata.JumioReturned {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: please wait for Jumio to finish processing your existing verification attempt before retrying."))
+		return
+	}
+
+	tStampNanos := int(time.Now().UnixNano())
+
+	jumioInternalReference := requestData.PublicKey + strconv.Itoa(tStampNanos)
+
+	userMetadata.JumioInternalReference = jumioInternalReference
+	userMetadata.JumioFinishedTime = 0
+	userMetadata.JumioReturned = false
+	if err = fes.putUserMetadataInGlobalState(userMetadata); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("JumioBegin: error putting jumio internal reference in global state: %v", err))
+		return
+	}
+
+	// CustomerInternalReference is Public Key + timestamp
+	// UserReference is just PublicKey
+	initData := &JumioInitRequest{
+		CustomerInternalReference: jumioInternalReference,
```

Comment:
Can we just use a UUID for this?

Reply:
we could but it significantly simplified things to have public key be a part of it when jumio flow finished stuff was happening in frontend instead of identity. I don't think it makes much of a difference either way now.